File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGSkin.js
#pragma strict
public enum OGStyleType {
	Basic,
	Hover,
	Active,
	Ticked,
	Thumb,
	Disabled
}

public enum OGWidgetType {
	NONE,
	Button,
	DropDown,
	Label,
	ListItem,
	PopUp,
	ProgressBar,
	ScrollView,
	Slider,
	SlicedSprite,
	Sprite,
	Tabs,
	TextField,
	TickBox
}

public class OGSlicedSpriteOffset {
	public var top : float;
	public var bottom : float;
	public var left : float;
	public var right : float;

	function OGSlicedSpriteOffset ( l : float, r : float, t : float, b : float ) {
		left = l;
		right = r;
		top = t;
		bottom = b;
	}

	public function get horizontal () : float {
		return left + right;
	}

	public function get vertical () : float {
		return top + bottom;
	}
}

public class OGTextStyle {
	public var fontIndex : int = 0;
	public var font : OGFont;
	public var fontSize : int = 12;
	public var fontColor : Color = Color.white;
	public var shadowSize : int = 0;
	public var shadowColor : Color = Color.black;
	public var alignment : TextAnchor;
	public var wordWrap : boolean = true;
	public var padding : RectOffset;
	public var lineHeight : float = 1.25;
	public var spacing : float = 1;
}

public class OGStyle {
	public var name : String = "New Style";
	public var coordinates : Rect = new Rect ( 0, 0, 32, 32 );
	public var border : OGSlicedSpriteOffset = new OGSlicedSpriteOffset ( 0, 0, 0, 0 );
	public var color : Color = Color.white;
	public var text : OGTextStyle;
}

public class OGStyleSet {
	public var basic : OGStyle;
	public var hover : OGStyle;
	public var active : OGStyle;
	public var ticked : OGStyle;
	public var thumb : OGStyle;
	public var disabled : OGStyle;

	public function Refresh ( skin : OGSkin ) {
		if ( !skin ) { return; }
		
		if ( basic ) { basic = skin.GetStyle ( basic.name ); }
		if ( hover ) { hover = skin.GetStyle ( hover.name ); }
		if ( active ) { active = skin.GetStyle ( active.name ); }
		if ( ticked ) { ticked = skin.GetStyle ( ticked.name ); }
		if ( thumb ) { thumb = skin.GetStyle ( thumb.name ); }
		if ( disabled ) { disabled = skin.GetStyle ( disabled.name ); }
	}

	public function GetStyle ( typ : OGStyleType ) {
		var result : OGStyle;
		
		switch ( typ ) {
			case OGStyleType.Basic:
				result = basic;
				break;

			case OGStyleType.Hover:
				result = hover;
				break;

			case OGStyleType.Active:
				result = active;
				break;

			case OGStyleType.Ticked:
				result = ticked;
				break;

			case OGStyleType.Thumb:
				result = thumb;
				break;

			case OGStyleType.Disabled:
				result = disabled;
				break;	
		}

		return result;
	}

	public function SetStyle ( typ : OGStyleType, stl : OGStyle ) {
		switch ( typ ) {
			case OGStyleType.Basic:
				basic = stl;
				break;

			case OGStyleType.Hover:
				hover = stl;
				break;

			case OGStyleType.Active:
				active = stl;
				break;

			case OGStyleType.Ticked:
				ticked = stl;
				break;

			case OGStyleType.Thumb:
				thumb = stl;
				break;
			
			case OGStyleType.Disabled:
				disabled = stl;
				break;	
		}
	}
}

public class OGDefault {
	public var widgetType : OGWidgetType;
	public var styleSet : OGStyleSet;

	function OGDefault ( t : OGWidgetType, s : OGStyleSet ) {
		widgetType = t;
		styleSet = s;
	}
}

public class OGSkin extends MonoBehaviour {
	public var atlas : Material;
	public var fonts : OGFont[] = new OGFont[0];
	public var fontShader : Shader;
	public var styles : OGStyle[] = new OGStyle[0];	
	public var defaults : OGDefault[] = new OGDefault[0];

	private static var widgetEnums : Dictionary.< System.Type, OGWidgetType > = new Dictionary.< System.Type, OGWidgetType > ();

	public static function GetWidgetEnum ( w : OGWidget ) : OGWidgetType {
		if ( widgetEnums.Count < 1 ) {
			widgetEnums.Add ( OGButton, OGWidgetType.Button );
			widgetEnums.Add ( OGDropDown, OGWidgetType.DropDown );
			widgetEnums.Add ( OGLabel, OGWidgetType.Label );
			widgetEnums.Add ( OGListItem, OGWidgetType.ListItem );
			widgetEnums.Add ( OGPopUp, OGWidgetType.PopUp );
			widgetEnums.Add ( OGProgressBar, OGWidgetType.ProgressBar );
			widgetEnums.Add ( OGScrollView, OGWidgetType.ScrollView );
			widgetEnums.Add ( OGSlider, OGWidgetType.Slider );
			widgetEnums.Add ( OGSlicedSprite, OGWidgetType.SlicedSprite );
			widgetEnums.Add ( OGSprite, OGWidgetType.Sprite );
			widgetEnums.Add ( OGTabs, OGWidgetType.Tabs );
			widgetEnums.Add ( OGTextField, OGWidgetType.TextField );
			widgetEnums.Add ( OGTickBox, OGWidgetType.TickBox );
		}

		if ( widgetEnums.ContainsKey ( w.GetType() ) ) {
			return widgetEnums [ w.GetType() ];
		} else {
			return OGWidgetType.NONE;
		}
	}

	public static function IsStyleUsed ( styleType : OGStyleType, widgetType : OGWidgetType ) : boolean {
		if ( widgetType == OGWidgetType.NONE ) { 
			return false;

		// All widgets use basic style
		} else if ( styleType == OGStyleType.Basic ) {
			return true;

		} else {
			var s : OGStyleType[];

			switch ( widgetType ) {
				case OGWidgetType.Button: s = [ OGStyleType.Disabled, OGStyleType.Active, OGStyleType.Hover, OGStyleType.Thumb ]; break;
				case OGWidgetType.DropDown: s = [ OGStyleType.Disabled, OGStyleType.Active, OGStyleType.Hover, OGStyleType.Ticked ]; break;
				case OGWidgetType.Label: s = [ OGStyleType.Disabled ]; break;
				case OGWidgetType.ListItem: s = [ OGStyleType.Disabled, OGStyleType.Hover, OGStyleType.Ticked ]; break;
				case OGWidgetType.PopUp: s = [ OGStyleType.Disabled, OGStyleType.Active, OGStyleType.Hover ]; break;
				case OGWidgetType.ProgressBar: s = [ OGStyleType.Thumb ]; break;
				case OGWidgetType.Slider: s = [ OGStyleType.Disabled, OGStyleType.Thumb ]; break;
				case OGWidgetType.ScrollView: s = [ OGStyleType.Thumb ]; break;
				case OGWidgetType.Tabs: s = [ OGStyleType.Disabled, OGStyleType.Active ]; break;
				case OGWidgetType.TextField: s = [ OGStyleType.Active, OGStyleType.Disabled ]; break;
				case OGWidgetType.TickBox: s = [ OGStyleType.Disabled, OGStyleType.Hover, OGStyleType.Ticked ]; break;
				default: return false;
			}

			for ( var i : int = 0; i < s.Length ; i++ ) {
				if ( styleType == s[i] ) {
					return true;
				}
			}
		}

		return false;
	}

	public function GetAllDefaults () : OGDefault[] {
		if ( defaults == null || defaults.Length < 1 ) {
			ResetDefaults ();
		}
	
		return defaults;
	}

	public function ResetDefaults () {
		defaults = [
			new OGDefault ( OGWidgetType.Button, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.DropDown, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Label, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ListItem, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.PopUp, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ProgressBar, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ScrollView, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Slider, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Tabs, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.TextField, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.TickBox, new OGStyleSet() )
		];
	}


	public function GetDefaultStyles ( w : OGWidget ) {
		for ( var d : OGDefault in GetAllDefaults() ) {
			if ( d.widgetType == w.ToEnum() ) {
				w.styles = new OGStyleSet();
				w.styles.basic = d.styleSet.basic;
				w.styles.active = d.styleSet.active;
				w.styles.hover = d.styleSet.hover;
				w.styles.thumb = d.styleSet.thumb;
				w.styles.disabled = d.styleSet.disabled;
				w.styles.ticked = d.styleSet.ticked;
			}
		}	
	}

	public function GetStyle ( n : String ) : OGStyle {
		for ( var s : OGStyle in styles ) {
			if ( s.name == n ) {
				return s;
			}
		}

		return null;
	}	

	public function GetAtlas () : Material {
		return Instantiate ( atlas );
	}
}

File: OpenGUI/Assets/Plugins/OpenGUI/Scripts/OGSkin.cs
using UnityEngine;
using System.Collections;
using System.Collections.Generic;

public enum OGStyleType {
	Basic,
	Hover,
	Active,
	Ticked,
	Thumb,
}

public enum OGWidgetType {
	NONE,
	Button,
	DropContainer,
	DropDown,
	Label,
	ListItem,
	PopUp,
	ProgressBar,
	ScrollView,
	Slider,
	SlicedSprite,
	Sprite,
	Tabs,
	TextField,
	TickBox
}

[System.Serializable]
public class OGSlicedSpriteOffset {
	public float top;
	public float bottom;
	public float left;
	public float right;

	public OGSlicedSpriteOffset ( float l, float r, float t, float b ) {
		left = l;
		right = r;
		top = t;
		bottom = b;
	}

	public float horizontal {
		get {
			return left + right;
		}
	}

	public float vertical {
		get {
			return top + bottom;
		}
	}
}

[System.Serializable]
public class OGTextStyle {
	public int fontIndex = 0;
	public OGFont font;
	public int fontSize = 12;
	public Color fontColor = Color.white;
	public int shadowSize = 0;
	public Color shadowColor = Color.black;
	public TextAnchor alignment;
	public bool wordWrap = true;
	public RectOffset padding;
	public float lineHeight = 1.25f;
	public float spacing = 1f;
}

[System.Serializable]
public class OGStyle {
	public string name = "New Style";
	public Rect coordinates = new Rect ( 0, 0, 32, 32 );
	public OGSlicedSpriteOffset border = new OGSlicedSpriteOffset ( 0, 0, 0, 0 );
	public Color color = Color.white;
	public OGTextStyle text;
}

[System.Serializable]
public class OGStyleSet {
	public OGStyle basic;
	public OGStyle hover;
	public OGStyle active;
	public OGStyle ticked;
	public OGStyle thumb;

	public void Refresh ( OGSkin skin ) {
		if ( !skin ) { return; }
		
		OGStyle newStyle;
		       
		if ( basic != null ) {
			newStyle = skin.GetStyle ( basic.name );
			if ( newStyle != null ) { basic = newStyle; }
		}
		
		if ( hover != null ) {
			newStyle = skin.GetStyle ( hover.name );
			if ( newStyle != null ) { hover = newStyle; }
		}
		
		if ( active != null ) {
			newStyle = skin.GetStyle ( active.name );
			if ( newStyle != null ) { active = newStyle; }
		}

		if ( ticked != null ) {
			newStyle = skin.GetStyle ( ticked.name );
			if ( newStyle != null ) { ticked = newStyle; }
		}
		
		if ( thumb != null ) {
			newStyle = skin.GetStyle ( thumb.name );
			if ( newStyle != null ) { thumb = newStyle; }
		}
	}

	public OGStyle GetStyle ( OGStyleType typ ) {
		OGStyle result = null;
		
		switch ( typ ) {
			case OGStyleType.Basic:
				result = basic;
				break;

			case OGStyleType.Hover:
				result = hover;
				break;

			case OGStyleType.Active:
				result = active;
				break;

			case OGStyleType.Ticked:
				result = ticked;
				break;

			case OGStyleType.Thumb:
				result = thumb;
				break;
		}

		return result;
	}

	public void SetStyle ( OGStyleType typ, OGStyle stl ) {
		switch ( typ ) {
			case OGStyleType.Basic:
				basic = stl;
				break;

			case OGStyleType.Hover:
				hover = stl;
				break;

			case OGStyleType.Active:
				active = stl;
				break;

			case OGStyleType.Ticked:
				ticked = stl;
				break;

			case OGStyleType.Thumb:
				thumb = stl;
				break;
		}
	}
}

[System.Serializable]
public class OGDefault {
	public OGWidgetType widgetType;
	public OGStyleSet styleSet;

	public OGDefault ( OGWidgetType t, OGStyleSet s ) {
		widgetType = t;
		styleSet = s;
	}
}

public class OGSkin : MonoBehaviour {
	public Material atlas;
	public OGFont[] fonts = new OGFont[0];
	public Shader fontShader;
	public OGStyle[] styles = new OGStyle[0];	
	public OGDefault[] defaults = new OGDefault[0];

	private static Dictionary< System.Type, OGWidgetType > widgetEnums = new Dictionary< System.Type, OGWidgetType > ();

	public static OGWidgetType GetWidgetEnum ( OGWidget w ) {
		if ( widgetEnums.Count < 1 ) {
			widgetEnums.Add ( typeof(OGButton), OGWidgetType.Button );
			widgetEnums.Add ( typeof(OGDropDown), OGWidgetType.DropDown );
			widgetEnums.Add ( typeof(OGDropContainer), OGWidgetType.DropContainer );
			widgetEnums.Add ( typeof(OGLabel), OGWidgetType.Label );
			widgetEnums.Add ( typeof(OGListItem), OGWidgetType.ListItem );
			widgetEnums.Add ( typeof(OGPopUp), OGWidgetType.PopUp );
			widgetEnums.Add ( typeof(OGProgressBar), OGWidgetType.ProgressBar );
			widgetEnums.Add ( typeof(OGScrollView), OGWidgetType.ScrollView );
			widgetEnums.Add ( typeof(OGSlider), OGWidgetType.Slider );
			widgetEnums.Add ( typeof(OGSlicedSprite), OGWidgetType.SlicedSprite );
			widgetEnums.Add ( typeof(OGSprite), OGWidgetType.Sprite );
			widgetEnums.Add ( typeof(OGTabs), OGWidgetType.Tabs );
			widgetEnums.Add ( typeof(OGTextField), OGWidgetType.TextField );
			widgetEnums.Add ( typeof(OGTickBox), OGWidgetType.TickBox );
		}

		if ( widgetEnums.ContainsKey ( w.GetType() ) ) {
			return widgetEnums [ w.GetType() ];
		} else {
			return OGWidgetType.NONE;
		}
	}

	public static bool IsStyleUsed ( OGStyleType styleType, OGWidgetType widgetType ) {
		if ( widgetType == OGWidgetType.NONE ) { 
			return false;

		// All widgets use basic style
		} else if ( styleType == OGStyleType.Basic ) {
			return true;

		} else {
			OGStyleType[] s = null;

			switch ( widgetType ) {
				case OGWidgetType.Button: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Hover, OGStyleType.Thumb }; break;
				case OGWidgetType.DropContainer: s = new OGStyleType[] { OGStyleType.Hover }; break;
				case OGWidgetType.DropDown: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Hover, OGStyleType.Thumb, OGStyleType.Ticked }; break;
				case OGWidgetType.ListItem: s = new OGStyleType[] { OGStyleType.Hover, OGStyleType.Ticked }; break;
				case OGWidgetType.PopUp: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Hover, OGStyleType.Thumb }; break;
				case OGWidgetType.ProgressBar: s = new OGStyleType[] { OGStyleType.Thumb }; break;
				case OGWidgetType.Slider: s = new OGStyleType[] { OGStyleType.Thumb }; break;
				case OGWidgetType.ScrollView: s = new OGStyleType[] { OGStyleType.Thumb }; break;
				case OGWidgetType.Tabs: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Thumb }; break;
				case OGWidgetType.TextField: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Thumb }; break;
				case OGWidgetType.TickBox: s = new OGStyleType[] { OGStyleType.Active, OGStyleType.Ticked }; break;
				default: return false;
			}

			for ( int i = 0; i < s.Length ; i++ ) {
				if ( styleType == s[i] ) {
					return true;
				}
			}
		}

		return false;
	}

	public OGDefault[] GetAllDefaults () {
		if ( defaults == null || defaults.Length < 1 ) {
			ResetDefaults ();
		}
	
		return defaults;
	}

	public void ResetDefaults () {
		defaults = new OGDefault[] {
			new OGDefault ( OGWidgetType.Button, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.DropDown, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.DropContainer, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Label, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ListItem, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.PopUp, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ProgressBar, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.ScrollView, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Slider, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.Tabs, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.TextField, new OGStyleSet() ),
			new OGDefault ( OGWidgetType.TickBox, new OGStyleSet() )
		};
	}

	public void ApplyDefaultStyles ( OGWidget w ) {
		foreach ( OGDefault d in GetAllDefaults() ) {
			if ( d.widgetType == w.ToEnum () ) {
				if ( w.styles == null ) {
					w.styles = new OGStyleSet ();
				}
			
				w.styles.basic = d.styleSet.basic;
				w.styles.active = d.styleSet.active;
				w.styles.hover = d.styleSet.hover;
				w.styles.thumb = d.styleSet.thumb;
				w.styles.ticked = d.styleSet.ticked;

				w.styles.Refresh ( this );
			
				return;
			}
		}	
	}
	
	public OGStyle GetStyle ( string n ) {
		foreach ( OGStyle s in styles ) {
			if ( s.name == n ) {
				return s;
			}
		}

		return null;
	}	

	public Material GetAtlas () {
		return (Material) Instantiate ( atlas );
	}
}
